csrc/optimization/consecutive_cast_pass.cpp

```diff
@@ -0,0 +1,101 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <optimization/consecutive_cast_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+void castOptimizationPass(Fusion* fusion) {
+  auto is_cast_op = [](Expr* expr) {
+    if (expr != nullptr && expr->isA<UnaryOp>()) {
+      auto op = expr->as<UnaryOp>();
+      if (op->getUnaryOpType() == UnaryOpType::Cast) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  // NOTE: not the most efficient pass
+  for (auto expr : fusion->exprs()) {
+    if (is_cast_op(expr)) {
+      bool mutated = false;
+      while (true) {
+        // in the loop, we just repetitively skip consecutive casts.
+        auto intermediate_cast = expr->input(0);
+        auto prev_expr = intermediate_cast->definition();
+        if (prev_expr == nullptr || !is_cast_op(prev_expr)) {
+          break;
+        }
+        // Note, if intermediate_cast:
+        //   is used by other none-cast operations; or
+        //   is a direct output from fusion
+        // we skip the casting
+        if (intermediate_cast->isFusionOutput() ||
+            !std::all_of(
+                intermediate_cast->uses().begin(),
+                intermediate_cast->uses().end(),
+                is_cast_op)) {
+          break;
```

Comment:
back out from folding the casts when there's other non-cast uses / intermediate is a fusion output.